### PR TITLE
feat: use `dtm init --download-only` to download specify plugins from flags

### PR DESCRIPTION
**Pull request by aFlyBird0**
Signed-off-by: Bird <aflybird0@gmail.com>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description
1. use `dtm init --download-only` to download specify plugins from flags.
2. use `--plugins` to specify plugins to download.
3. use `--all` to download all plugins.
4. use `--os` and `--arch` to specify os and arch, the default is os and arch of the machine which executes dtm.

## Related Issues
The first task of #1084.

(We can use quick-start script to download `dtm` and use `dtm init --all --os=DestinyOS --arch=DestinyArch -d=PLUGIN_DIR` to download plugins.

## New Behavior (screenshots if needed)

Comment out this code and test:

image

image

image

image

### CODE REVIEW

**Comment by daniel-hutao** [comment]
@aFlyBird0 Hi bird, I have a small question here. Why do we need `--all` flag here?

**Comment by aFlyBird0** [comment]
> @aFlyBird0 Hi bird, I have a small question here. Why do we need `--all` flag here?

We need to add a flag to download all plugins so that users could use `dtm` offline. If `--all` flag is set, config file can be empty.

If not, users must create a config file and execute `dtm init`.

**Comment by daniel-hutao** [comment]
> > @aFlyBird0 Hi bird, I have a small question here. Why do we need `--all` flag here?
> 
> We need to add a flag to download all plugins so that users could use `dtm` offline. If `--all` flag is set, config file can be empty.
> 
> If not, users must create a config file and execute `dtm init`.

That is, if a user needs to use 3 plugins only, then we suggest he/she download all plugins.

Maybe there's nobody who needs to use all plugins. Do most users need to download just 1-5? plugins, but dtm has 28 plugins(maybe more than 100 plugins next year).

So my question is: what if a user needs to download 5/100 plugins, and what should he/she do?

**Comment by daniel-hutao** [comment]
How about we use `dtm init --download-only --plugins="jenkins, harbor, gitlab"` to download plugins? `--all` flag equals false by default.

**Comment by aFlyBird0** [comment]
> How about we use `dtm init --download-only --plugins="jenkins, harbor, gitlab"` to download plugins? `--all` flag equals false by default.

Cool, I agree with you! 🚀

File: cmd/devstream/init.go [hunk 1]
```diff
@@ -20,35 +24,119 @@ var initCMD = &cobra.Command{
...
+		log.Fatalf("Dev version plugins can't be downloaded from the remote plugin repo; please run `make build-plugin.PLUGIN_NAME` to build them locally.")
+	}
+
+	var (
+		realPluginDir string
+		tools         []configmanager.Tool
+		err           error
+	)
+
+	switch downloadOnly {
```
Comment (daniel-hutao) [inline cmd/devstream/init.go:46]:
```go
	switch downloadOnly {
 	// download plugins according to the config file
 	case false:
 		tools, realPluginDir, err = GetPluginsAndPluginDirFromConfig()
 	// download plugins from flags
 	case true:
 		tools, realPluginDir, err = GetPluginsAndPluginDirFromFlags()
 	}
```

vs

```go
if downloadOnly {
    tools, realPluginDir, err = GetPluginsAndPluginDirFromFlags()
} else {
    tools, realPluginDir, err = GetPluginsAndPluginDirFromConfig()
}
```

I think the "if" version is more clear here.

File: cmd/devstream/init.go [hunk 2]
```diff
@@ -20,35 +24,119 @@ var initCMD = &cobra.Command{
...
 }
 
 func init() {
+	// flags for init from config file
 	initCMD.Flags().StringVarP(&configFile, configFlagName, "f", "config.yaml", "config file")
 	initCMD.Flags().StringVarP(&pluginDir, pluginDirFlagName, "d", "", "plugins directory")
 
+	// downloading specific plugins from flags
+	initCMD.Flags().BoolVar(&downloadOnly, "download-only", false, "download plugins only")
+	initCMD.Flags().StringSliceVarP(&pluginsToDownload, "plugins", "p", []string{}, "plugins to download")
```
Comment (daniel-hutao) [inline cmd/devstream/init.go:135]: "the plugins to be downloaded"?

**Review by daniel-hutao — CHANGES_REQUESTED** [review]

File: cmd/devstream/init.go (hunk 1 above)
Comment (aFlyBird0) [reply cmd/devstream/init.go:46]: That sounds great.

**Review by daniel-hutao — APPROVED** [review]

**Pull request merged**